uaclient/contract.py

```diff
@@ -171,6 +173,106 @@ def report_machine_activity(self):
+        headers.update({"Authorization": "Bearer {}".format(magic_id)})
+        url = API_V1_MAGIC_ATTACH
+
+        try:
+            response, _ = self.request_url(url, headers=headers)
+        except exceptions.UrlError as e:
+            if isinstance(e, exceptions.ContractAPIError):
+                if hasattr(e, "code"):
+                    if e.code == 400:
+                        raise exceptions.MagicAttachExpiredID(
+                            magic_id=magic_id
+                        )
+                    elif e.code == 401:
+                        raise exceptions.MagicAttachInvalidID(
+                            magic_id=magic_id
+                        )
+                    elif e.code == 403:
+                        raise exceptions.MagicAttachInvalidUser()
+                    elif e.code == 404:
+                        raise exceptions.MagicAttachIDNotResolved(
+                            magic_id=magic_id
+                        )
+                raise e
+
+            logging.exception(str(e))
+            raise exceptions.UserFacingError(
+                msg=messages.CONNECTIVITY_ERROR.msg,
+                msg_code=messages.CONNECTIVITY_ERROR.name,
+            )
+
+        return response
+
+    def request_magic_attach_id(self, email: str) -> Dict[str, Any]:
+        """Request a magic attach id for the user.
+
+        This id should allow the user to log into the pro portal
+        and associate the machine with the id when choosing a contract
+        to attach to.
+        """
+        headers = self.headers()
+        url = API_V1_MAGIC_ATTACH
+
+        try:
+            kwargs = {
+                "headers": headers,
+                "method": "POST",
+                "data": {"email": email},
+            }
+            response, _ = self.request_url(url, **kwargs)
+        except exceptions.UrlError as e:
+            logging.exception(str(e))
+            raise exceptions.UserFacingError(
+                msg=messages.CONNECTIVITY_ERROR.msg,
+                msg_code=messages.CONNECTIVITY_ERROR.name,
+            )
+
+        return response
+
+    def delete_magic_attach_id(self, magic_id: str):
+        """Delete a magic attach id for the user.
+
+        This id should allow the user to log into the pro portal
+        and associate the machine with the id when choosing a contract
+        to attach to.
+        """
+        headers = self.headers()
+        headers.update({"Authorization": "Bearer {}".format(magic_id)})
+        url = API_V1_MAGIC_ATTACH
+
+        try:
+            kwargs = {
+                "headers": headers,
+                "method": "DELETE",
+            }
+            self.request_url(url, **kwargs)
+        except exceptions.UrlError as e:
+            if isinstance(e, exceptions.ContractAPIError):
+                if hasattr(e, "code"):
+                    if e.code == 400:
+                        raise exceptions.MagicAttachIDExpiredOrActivated(
```

Comment:
So, you should handle it, but it only means the token is already activated.

If it's expired, you will get a 401.

Reply:
Just to double check, you mean that we should not error out to the user when trying to revoke an activated token, just state that it is already activated, right ?

Reply:
For all endpoints that accept a "Bearer <magic-attach-token>" kind of auth, `401` always means one of: not found, invalid format (e.g., someone said the token is "abc123") or expired already.

The message from ua-contracts will not distinguish between these cases, and it will just say "unauthorized".

I'd suggest: `exceptions.MagicAttachTokenError` with the following message to the user:
> The magic attach token is invalid, has expired or never existed.